File: src/Game.fs
module ThreesAI.Game

open System
open Controls

let private random = Random ()
    
type State = {
    Board: Board
    NextTile: Tile
    Deck: Deck
    BonusDeck: Deck
    Score: int
    GameOver: bool
}
    
let private getPlacementSpots direction state =
    let edge =
       match direction with
       | Up    -> [for x in [0..3] do (x, 3), state.Board.[x, 3]]
       | Down  -> [for x in [0..3] do (x, 0), state.Board.[x, 0]]
       | Left  -> [for y in [0..3] do (3, y), state.Board.[3, y]]
       | Right -> [for y in [0..3] do (0, y), state.Board.[0, y]]
    
    edge |> List.filter (fun (_, space) -> space = Empty)
    
let private setBonusDeck state =
    let bestTile = Board.max state.Board |> Board.getTileValue
    if bestTile >= 48 then
        if List.length state.BonusDeck = 0 || bestTile / 8 <> (List.head state.BonusDeck |> Board.getTileValue) then
            { state with BonusDeck = Tile (bestTile / 8) :: state.BonusDeck }
        else
            state
    else
        state
    
let private getNextTile state =
    if List.length state.BonusDeck > 0 && random.Next(0, 21) = 0 then
        { state with NextTile = state.BonusDeck.[random.Next (0, (List.length state.BonusDeck) - 1)] }
    else
        let deck = if List.length state.Deck > 0 then state.Deck else Deck.create ()
        { state with NextTile = List.head deck; Deck = List.tail deck }
    
let private placeTile direction state =
    let spots = getPlacementSpots direction state
    let x, y = fst <| spots.[random.Next(0, (List.length spots) - 1)]
    
    // Lazy, but Array2D is mutable so shrug
    let newBoard = Array2D.copy state.Board
    newBoard.[x, y] <- state.NextTile
    
    { getNextTile state with Board = newBoard; }

let didShiftOccur before after =
    let beforeFlat = Seq.cast<Tile> before
    let afterFlat = Seq.cast<Tile> after
    
    Seq.contains false <| Seq.map2 (fun a b -> a = b) beforeFlat afterFlat

let private canTileMerge (board: Board) x y tile =
    let adjacent =
        [ Board.tryGet (x - 1, y) board
          Board.tryGet (x + 1, y) board
          Board.tryGet (x, y - 1) board
          Board.tryGet (x, y + 1) board ] |> List.filter (fun result -> result <> None) |> List.map unwrap

    let mergeResults = adjacent |> List.map (Board.canMerge tile)
    List.contains true mergeResults

let private detectGameOver state =
    let canTilesMerge = state.Board |> Array2D.mapi (canTileMerge state.Board) |> Seq.cast<bool> |> Seq.contains true
    if canTilesMerge then
        state
    else
        { state with GameOver = true }

let calculateScore state =
    let scoreTile totalScore tile =
        match tile with
        | Empty  -> totalScore
        | Tile 1 -> totalScore
        | Tile 2 -> totalScore
        | Tile t -> totalScore + (int <| 3.0 ** (Math.Log2(float t / 3.0) + 1.0))
    
    { state with Score = Seq.cast<Tile> state.Board |> Seq.fold scoreTile 0 }
    
let shift direction state =
    let shifted = { state with Board = state.Board |> Board.shift direction }
    if didShiftOccur state.Board shifted.Board then
        setBonusDeck shifted
        |> placeTile direction
        |> calculateScore
        |> detectGameOver
    else
        state

let create () =
    let deck = Deck.create ()
    let board = Board.empty ()
    
    // I could do any number of unoptimized non-mutable methods
    // but this works a lot better
    let rec placementLoop tilesPlaced (deck: Deck) =
        if tilesPlaced = 9 then
            deck
        else
            
        let rx, ry = (random.Next (0, 4), random.Next (0, 4))
        if board.[rx, ry] = Empty then
            let next = List.head deck
            board.[rx, ry] <- next
            placementLoop (tilesPlaced + 1) (List.tail deck)
        else
            placementLoop tilesPlaced deck
    
    let newDeck = placementLoop 0 deck
    { Board = board
      NextTile = List.head newDeck
      Deck = List.tail newDeck
      BonusDeck = []; Score = 0
      GameOver = false } |> calculateScore

File: ai/game.py
# Port of F# code to python

from functools import reduce
from math import log2
from typing import List, cast
import numpy as np
from random import shuffle, randrange, choice

import board

class Threes():
    board: np.ndarray
    next_tile: int
    deck: List[int]
    bonus_deck: List[int]
    score: int
    game_over: bool

    def getPlacementSpots(self, direction: np.ndarray) -> List[tuple[tuple[int, int], int]]:
        edge = []
        
        if direction == board.Direction.UP:
            edge = [((3, x), self.board[3, x]) for x in range(4)]
        elif direction == board.Direction.DOWN:
            edge = [((0, x), self.board[0, x]) for x in range(4)]
        elif direction == board.Direction.LEFT:
            edge = [((y, 3), self.board[y, 3]) for y in range(4)]
        else:
            edge = [((y, 0), self.board[y, 0]) for y in range(4)]
        
        return list(filter(lambda e: e[1] == 0, edge))
    
    def setBonusDeck(self):
        bestTile = np.max(self.board)

        if bestTile < 48:
            return
        
        if len(self.bonus_deck) == 0 or bestTile / 8 != self.bonus_deck[-1]:
            self.bonus_deck.append(bestTile / 8)

    def getNextTIle(self):
        if len(self.bonus_deck) > 0 and randrange(0, 21) == 0:
            self.next_tile = choice(self.bonus_deck)
        else:
            if len(self.deck) == 0:
                self.createDeck()
            
            self.next_tile = self.deck.pop()
    
    def createDeck(self):
        self.deck = [1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3]
        shuffle(self.deck)
    
    def placeTile(self, direction):
        spots = self.getPlacementSpots(direction)
        x, y = choice(spots)[0]

        self.board[x, y] = self.next_tile
    
    def didShiftOccur(self, before: np.ndarray, after: np.ndarray) -> bool:
        beforeFlat = before.flatten()
        afterFlat = after.flatten()

        return False in [a == b for a, b in zip(beforeFlat, afterFlat)]
    
    def canTileMerge(self, x, y, tile):
        adjacent = [
            board.tryGet(x - 1, y, self.board),
            board.tryGet(x + 1, y, self.board),
            board.tryGet(x, y - 1, self.board),
            board.tryGet(x, y + 1, self.board) 
        ]

        adjacent = list(filter(lambda result: result != None, adjacent))
        mergeResults = list(map(lambda t: board.canMerge(tile, cast(int, t)), adjacent))

        return True in mergeResults
    
    def detectGameOver(self):
        canTilesMerge = False
        for x in range(4):
            for y in range(4):
                if self.canTileMerge(x, y, self.board[x, y]):
                    canTilesMerge = True
                    break
            
            if canTilesMerge:
                break
        
        self.game_over = not canTilesMerge
    
    def getTileScore(self, tile: int):
        if tile >= 3:
            return int(3 ** (log2(tile / 3) + 1))
        else:
            return 0

    def calculateScore(self):
        flat = self.board.flatten()

        self.score = 0
        for tile in flat:
            self.score += self.getTileScore(tile)
    
    def shift(self, direction):
        shifted = board.shift(direction, self.board)
        if self.didShiftOccur(self.board, shifted):
            self.board = shifted
            self.setBonusDeck()
            self.placeTile(direction)
            self.calculateScore()
            self.detectGameOver()

    def reset(self):
        self.createDeck()
        self.board = board.empty()

        tilesPlaced = 0

        while tilesPlaced < 9:
            rx, ry = (randrange(0, 4), randrange(0, 4))
            if self.board[rx, ry] == 0:
                self.board[rx, ry] = self.deck.pop()
                tilesPlaced += 1
        
        self.next_tile = self.deck.pop()
        self.bonus_deck = []
        self.calculateScore()
        self.game_over = False

    def __init__(self) -> None:
        self.reset()